kafka-0.10.x/src/main/scala/monix/kafka/KafkaProducerSink.scala

```diff
@@ -39,29 +40,26 @@ final class KafkaProducerSink[K, V] private (
 
   require(parallelism >= 1, "parallelism >= 1")
 
-  def createSubscriber(cb: Callback[Throwable, Unit], s: Scheduler) = {
+  def createSubscriber(
+    cb: Callback[Throwable, Unit],
+    s: Scheduler): (Subscriber[Seq[ProducerRecord[K, V]]], AssignableCancelable.Multi) = {
     val out = new Subscriber[Seq[ProducerRecord[K, V]]] { self =>
-      implicit val scheduler = s
+      implicit val scheduler: Scheduler = s
       private[this] val p = producer.memoize
       private[this] var isActive = true
 
-      private def sendAll(batch: Seq[ProducerRecord[K, V]]): Seq[Task[Option[RecordMetadata]]] =
-        for (record <- batch) yield {
-          try p.value.send(record)
-          catch { case NonFatal(ex) => Task.raiseError(ex) }
-        }
-
       def onNext(list: Seq[ProducerRecord[K, V]]): Future[Ack] =
         self.synchronized {
           if (!isActive) Stop
           else {
             val sendTask: Task[Seq[Option[RecordMetadata]]] =
               if (parallelism == 1)
-                Task.sequence(sendAll(list))
+                Task.traverse(list)(p.value().send(_))
               else {
-                val batches = list.sliding(parallelism, parallelism)
-                val tasks = for (b <- batches) yield Task.gather(sendAll(b))
-                Task.sequence(tasks.toList).map(_.flatten)
+                for {
+                  s <- Semaphore[Task](parallelism)
```

Comment:
We could reuse this semaphore instead of creating it each `onNext`

Reply:
Added memoization